File: BggExt/Models/BoardGame.cs
using Microsoft.EntityFrameworkCore;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;

namespace BggExt.Models;

public class BoardGame
{
    public int Id { get; set; }

    [Required()]
    public string Name { get; set; } = default!;

    public string? Description { get; set; }

    public int YearPublished { get; set; }

    public int MinPlayers { get; set; }

    public int MaxPlayers { get; set; }

    public int PlayingTimeMinutes { get; set; }

    public int MinPlayTimeMinutes { get; set; }

    public int MaxPlayTimeMinutes { get; set; }

    public int MinAge { get; set; }

    public double AverageWeight { get; set; }

    public ICollection<Mechanic>? Mechanics { get; set; }

    public ICollection<Category>? Categories { get; set; }

    public ICollection<Family>? Families { get; set; }

    public Image? Thumbnail { get; set; }

    public Image? Image { get; set; }


    public static BoardGame ExampleBoardGame() =>
        new()
        {
            Name = "Example Board Game",
            Description = "This is a random board game",
            YearPublished = 2010,
            MinPlayers = 1,
            MaxPlayers = 4,
            PlayingTimeMinutes = 30,
            MinPlayTimeMinutes = 15,
            MaxPlayTimeMinutes = 60,
            MinAge = 10,
            AverageWeight = 2.5,
            Categories = new List<Category>
            {
                new() { Name = "Category 1" },
                new() { Name = "Category 2" },
                new() { Name = "Category 3" }
            },
            Mechanics = new List<Mechanic>
            {
                new() { Name = "Mechanic 1" },
                new() { Name = "Mechanic 2" },
                new() { Name = "Mechanic 3" }
            },
            Families = new List<Family>
            {
                new() { Name = "Family 1" },
                new() { Name = "Family 2" },
                new() { Name = "Family 3" }
            },
            Thumbnail = new Image
            {
                ImageData = new byte[] { 0x01, 0x02, 0x03, 0x04 }
            },
            Image = new Image
            {
                ImageData = new byte[] { 0x01, 0x02, 0x03, 0x04 }
            }
        };
}

File: BggExt/ClientApp/src/components/BoardGame.js
import React, { useEffect, useState } from 'react';
import PropTypes from "prop-types";
import useToken from "../useToken";

// type BoardGame = {
//     id: number;
//     name: string;
//     description: string;
//     yearPublished: number;
//     minPlayers: number;
//     maxPlayers: number;
//     playingTimeMinutes: number;
//     minPlayTimeMinutes: number;
//     maxPlayTimeMinutes: number;
//     minAge: number;
//     averageWeight: number;
//     mechanics: Mechanic[];
//     categories: Category[];
//     families: Family[];
//     thumbnail?: Image;
//     image?: Image;
// };

// type Mechanic = {
//     id: number;
//     name: string;
//     boardGames: BoardGame[] | null;
// };

// type Category = {
//     id: number;
//     name: string;
//     boardGames: BoardGame[] | null;
// };

// type Family = {
//     id: number;
//     name: string;
//     boardGames: BoardGame[] | null;
// };

// type Image = {
//     id: number;
//     imageData: string;
// };
export default function BoardGame({ boardGameId }) {
    const { token, setToken } = useToken();
    console.log(token);
    const accessToken = token?.accessToken;
    console.log(`accessToken: ${accessToken}`);
    // fetch games and map to type
    const [game, setGames] = useState();
    const url = `BoardGame/${boardGameId}`;
    console.log(url);
    useEffect(() => {
        fetch(url, {
            method: 'GET',
            headers: new Headers({
                'Authorization': `Bearer ${accessToken}`
            }),
        })
            .then(response => response.json())
            .then(data => {
                setGames(data);
            });
    }, []);

    const renderArray = (array) => {
        console.log("renderArray");
        console.log(array);
        // Filter out null or undefined entries
        const filteredArray = array.filter((item) => item);
        return filteredArray.map((item) => item.name).join(', ');
    };

    if (game == null) {
        return null;
    }
    console.log("game");
    console.log(game);
    return (
        <table className='table table-striped' aria-labelledby='tabelLabel'>
            <thead>
                <tr>
                    <th>Name</th>
                    <th>Year Published</th>
                    <th>Min Players</th>
                    <th>Max Players</th>
                    <th>Playing Time</th>
                    <th>Min Play Time</th>
                    <th>Max Play Time</th>
                    <th>Min Age</th>
                    <th>Average Weight</th>
                    <th>Mechanics</th>
                    <th>Categories</th>
                    <th>Families</th>
                </tr>
            </thead>
            <tbody>
                <tr key={game.id}>
                    <td>{game.name}</td>
                    <td>{game.yearPublished}</td>
                    <td>{game.minPlayers}</td>
                    <td>{game.maxPlayers}</td>
                    <td>{game.playingTimeMinutes}</td>
                    <td>{game.minPlayTimeMinutes}</td>
                    <td>{game.maxPlayTimeMinutes}</td>
                    <td>{game.minAge}</td>
                    <td>{game.averageWeight}</td>
                    <td>{renderArray(game.mechanics)}</td>
                    <td>{renderArray(game.categories)}</td>
                    <td>{renderArray(game.families)}</td>
                </tr>
            </tbody>
        </table>
    );
};

BoardGame.propTypes = {
    boardGameId: PropTypes.func.isRequired
};
